### PR TITLE
Add function for setting max query execution time to core

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Add function for setting max query execution time to core

Before
----------------------------------------
@twomice has figured out how to leverage mysql / mariaDB max execution variable in an extension but it is not available to core / other extensions

https://github.com/JoineryHQ/com.joineryhq.dupmon/blob/master/CRM/Dupmon/Util.php#L126-L131

After
----------------------------------------
New Core functions 
`CRM_Core_DAO::getMaxExecutionTime()`
`CRM_Core_DAO::setMaxExecutionTime()`

With recommended usage being
```
$autoClean = CRM_Utils_AutoClean::swapMaxExecutionTime(800);
$autoClean->cleanup();
```

Technical Details
----------------------------------------
We don't know if there are any non-handled edge cases - hopefully tests can try a few sql versions.... 

However, this currently has no core callers outside of the unit test so we can add it cautiously.

I propose we follow up by adding a core caller in `hook_civicrm_findExistingDuplicates()` - this core code currently has no users as it is only called when the new hidden, installed on upgrade and new installs `legacydedupefinder` extension is enabled. This is allowing is to roll out less tested but more performant dedupe methodology.

We would need to determine a core setting to replace the one in https://github.com/JoineryHQ/com.joineryhq.dupmon/blob/master/CRM/Dupmon/Util.php#L194C6-L194C27 and possibly think about managing the migration with @twomice - either a version check or cutting a new extension release for 6.1 would do it

Comments
----------------------------------------

### CODE REVIEW

**Comment by totten** [comment]
Ooh, neat idea.

**Comment by twomice** [comment]
I've unsubscribed, but please ping me again if you think I can help.

**Comment by twomice** [comment]
Cool doings, though, @eileenmcnaughton !

**Comment by demeritcowboy** [comment]
I tried this out (mariadb). It seems to work (DB Error thrown when query exceeds the time limit). One thing that my read of the docblock was off and I struggled for a bit trying to figure out that you have to assign the autoclean to a variable. Maybe the order of the wording in the docblock, e.g. change

```
   * avoids the situation where someone changes the above line to
   * the self-reverting (because `__destruct` is implicitly called)
   * `CRM_Utils_AutoClean::swapMaxExecutionTime(800);`
```

to something like

```
   * avoids the situation where someone just calls
   * `CRM_Utils_AutoClean::swapMaxExecutionTime(800);`
   * without assigning it to a variable (because `__destruct` is implicitly called)
```

File: CRM/Core/DAO.php [hunk 1]
```diff
@@ -219,6 +219,72 @@ public function getLog() {
...
+   *
+   * https://mariadb.com/kb/en/aborting-statements/
+   *
+   * @return int
+   * @throws \Civi\Core\Exception\DBQueryException
+   */
+  public static function getMaxExecutionTime(): int {
+    $version = CRM_Utils_SQL::getDatabaseVersion();
+    if (stripos($version, 'mariadb') !== FALSE) {
+      $originalSql = 'SHOW VARIABLES LIKE "%MAX_STATEMENT_TIME%"';
```
Comment (demeritcowboy) [inline CRM/Core/DAO.php:274]: Does this need percent signs? Similarly line 280.

Comment (eileenmcnaughton) [reply CRM/Core/DAO.php:274]: @demeritcowboy I thought that - I took it from @twomice - Allan can you confirm?

**Comment by eileenmcnaughton** [comment]
@demeritcowboy OK - I made the comment change you suggested & removed the `%` signs since twomice seems to concur

**Comment by demeritcowboy** [comment]
Sounds goo... `DB Error: timeout`
